Consul Upstream Load Balancing lab (#244)
hashicorp/katakoda: consul-upstream-load-balancing/step3.md

> The default load balancing policy, _random_, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.

Author: Given the fact we have checks and that LB is made on an healthy pool the statement "best approach" is still valid if applied to round robin.

> The default load balancing policy, *round_robin*, is usually the best approach in scenarios where service nodes capacity is homogeneous and when you can rely on a health check policy that prevents failed nodes to be used for the scheduling. If you configure health checks for your services you'll automatically benefit from them during the load balancing process.

> The default load balancing policy, _random_, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.

Reviewer: Hmm cool, then we just need to replace the word `random`

In scenarios where the different instances might be subject to substantial differences in terms of workload there are better approaches.

> Using the *least_request* policy permits Envoy to resolve requests using information on instance load level and select the one with the lowest load.

Reviewer 2: ```suggestion
Using the *least_request* policy permits Envoy sidecar proxies to resolve requests using information on instance load level and select the one with the lowest load.
```

### Configure least_request load balancing policy

The lab provides an example configuration file, `least-req-resolver.hcl`{{open}} for *least_request* policy.

```
Kind           = "service-resolver"
Name           = "backend"
LoadBalancer = {
  Policy = "least_request"
  LeastRequestConfig = {
    ChoiceCount = "2"
  }
}
```

> This configuration creates a `service-resolver` configuration, for the service `backend` that for every request picks 2 (as expressed by `ChoiceCount`) random instances of the service and resolves to the one with the least amount of load.

Reviewer 2: ```suggestion
This configuration creates a `service-resolver` load balancing policy, for every request to the `backend` service 2 (as expressed by `ChoiceCount`) random instances of that service are selected and the traffic is routed to the one with the least amount of load.
```

You can apply the policy using the `consul config` command.

`docker exec server consul config write /etc/consul.d/least-req-resolver.hcl`{{execute}}

Example output:

```
Config entry written: service-resolver/backend
```

### Verify the policy is applied

> Once the policy is in place you can test it using the `curl` command and applying the `x-user-id` header to the request:

Reviewer 2: ```suggestion
Once the policy is in place, you can test it using the `curl` command and applying the `x-user-id` header to the request:
```

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}

Example output:

```
{
  "name": "main",
  "uri": "/",
  "type": "HTTP",
  "ip_addresses": [
    "172.18.0.4"
  ],
  "start_time": "2020-10-01T16:25:47.950151",
  "end_time": "2020-10-01T16:25:47.950581",
  "duration": "420.066µs",
  "body": "Hello World",
  "code": 200
}
```

> Executing the command multiple times you can verify that, despite you are still applying the header that was used by the previous policy the request gets served by different instances of the service.

Reviewer 2: ```suggestion
Execute the command multiple times to verify that, despite the user id in the header, the request gets served by different instances of the service.

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}
```

> The default load balancing policy, _random_, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.

Author: ```suggestion
The default load balancing policy, *round_robin*, is usually the best approach in scenarios where requests are homogeneous and the system is over-provisioned.
```
